### DISCUSSION TITLE
is variable feature from NodeBipartiteObs  sorted by unique index?

### DISCUSSION THREAD

**Discussion by th-yoon** [ds4dm/ecole#236] [Questions] [answered]
Hi, I have a question about  variable features extracted from NodeBipartiteObs module.
I see that it gets columns as in [this line ](https://github.com/ds4dm/ecole/blob/6cc08bb2a8dfff1d15b3e907e67a6ac22f6d9ee9/libecole/src/observation/nodebipartite.cpp#L174), and it depends on `SCIPgetLPCols()` function in SCIP C code. 
Does it ensure that the variable features are sorted row-wise based on unique index of each variable?
Thanks.

**Comment by gasse** [comment, marked as answer]
Hi @th-yoon ,

The variable features in `NodeBipartiteObs` indeed are extracted for all variables in the current node's LP (obtained from `SCIPgetLPCols()`). They are also registered in the same order, and thus can be indexed using the column position (`var.getCol().getLPPos()`) of the variables. So, if you have a list of variables and you want to get their associated features in `NodeBipartiteObs`, you can use something like this:
```python
import numpy as np

obs = ...  # NodeBipartiteObs

m = model.as_pyscipopt()
vars = m.getVars(transformed=True)
vars_in_lp = [var for var in vars if var.isInLP()]
lppos_vars_in_lp = np.asarray([var.getCol().getLPPos() for var in vars_in_lp])
feats_vars_in_lp = obs.column_features[lppos_vars_in_lp]
```

See also this post which is very related:
https://github.com/ds4dm/ml4co-competition/discussions/24#discussioncomment-1027046

Hope that helps,

Best,
Maxime

**Reply by th-yoon** [reply]
Thanks @gasse , it helped a lot!
